### DISCUSSION TITLE
[Ansible] Can't find Ansible with self-hosted agent running Ubuntu 22.04 image

### DISCUSSION THREAD

**Discussion by Kor-Hal** [actions/virtual-environments#7823] [Installed software]
Hello everyone,

I have an issue regarding Ansible and the Ubuntu 22.04 Runner image.
When using the Ansible@0 task in Azure Devops on a self-hosted agent using the latest Ubuntu 22.04 image, I get the following error :
##[error]Could not locate Ansible. Either it is not installed or its path is not present on ENVIRONMENT PATH variable

To make sure it wasn't dependent on my setup, I asked another person on a Discord to test with their setup and they got the same error.
I also tested with the official Azure Pipelines and it works there (using ubuntu-latest vmImage).

Does anybody have any clue about why this happens and how to solve it ?

Thanks for your help.

**Comment by Kor-Hal** [comment]
After further investigation, it seems that the path value isn't the same between hosted agents and self hosted ones.

Hosted :
/home/vsts/.local/bin:/opt/pipx_bin:/home/vsts/.cargo/bin:/home/vsts/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/vsts/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin

Self-hosted :
/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin

Ansible living in /opt/pipx_bin

Could be some missing step in building the image but I can't find where :(

**Reply by vpolikarpov-akvelon** [reply]
Hey @Kor-Hal. As `PATH` is an environment variable that lives only in runtime it cannot be set during image generation. Instead it's value is saved to `/etc/environment` file. For example, look here: https://github.com/actions/runner-images/blob/main/images/linux/scripts/installers/python.sh#L22

When runner is started, this file is read and applied. You can try doing the same if you want your environment to be the same as on GitHub-hosted runners.

**Reply by Kor-Hal** [reply]
Hello and thank you for your answer :)

As a foreword, I found a Github issue with exactly what I'm facing : https://github.com/microsoft/azure-pipelines-agent/issues/3461

I had found out that the value was saved in /etc/environment.
What I don't understand (or know how to do) is how to "read and apply" that file when the VMSS is popping instances.

So far I've tried to add a custom script (found here : https://github.com/microsoft/azure-pipelines-agent/issues/3461#issuecomment-1250121953) as an extension but it doesn't seem to help.

I'm building and deploying my VMSS with an approach very similar to this one : https://github.com/YannickRe/azuredevops-buildagents

Can you please tell/teach me how to have the runner read and apply the /etc/environment file when an instance of the VMSS is started ?
Thanks a lot for your help !

**Reply by vpolikarpov-akvelon** [reply]
There is an environment variable you can set to execute custom script before each job: https://docs.github.com/en/enterprise-cloud@latest/actions/hosting-your-own-runners/managing-self-hosted-runners/running-scripts-before-or-after-a-job. But I'm not sure whether variables set in hook script are available during actual workload execution.

Also, you can try copying content of `/etc/environment` file directly to `.env` file in runner root directory either manually or on VM provisioning.

**Reply by Kor-Hal** [reply]
Copying the content into .env (and .path) worked :) Thanks.

I wish this startup script would be provided as part of the image and/or agent extension :(

**Reply by vpolikarpov-akvelon** [reply]
This repository focuses on images used for GitHub-hosted runners, which are designed to align with the unique aspects of the GitHub infrastructure. As a result, these images do not include the actual runner software, nor do they come pre-configured with any runner settings. It up to you to install and configure the runner software according to your specific requirements.

**Reply by Kor-Hal** [reply]
I understand (and agree with) that.

I wish the Azure Devops agent extension would include a copy of /etc/environment into /agent/.path or /agent/.env.
But as stated in the issue I previously linked, it doesn't seem to be the plan :(

When I used the runner images, I saw the readmes and assumed the tools would be available as soon as the packer build command was done, which isn't the case for Ansible and other ones (and can't be because, as you mentionned, it depends on specific requirements that can't be known at the time the packer build command is run).

Anyway, I finally got everything working with the script provided in the issue linked above, so everything is fine now :)
Thanks for your answers and have a nice day.
